# Revisão de código 

Olá! 

Meu nome é Suelen. Estou feliz em revisar seu projeto hoje!

Quando vejo um erro pela primeira vez, apenas aponto. Deixarei você encontrá-lo e corrigi-lo sozinho. Além disso, ao longo do texto, farei algumas observações sobre melhorias no código e também farei comentários sobre suas percepções sobre o assunto. Mas se você ainda não consegue lidar com essa tarefa, darei uma dica mais precisa na próxima iteração e também alguns exemplos práticos. Estarei aberta a feedbacks e discussões sobre o tema.

Você pode encontrar meus comentários em caixas verdes, amarelas ou vermelhas como estas:


<div class="alert alert-block alert-success">
<b>Comentário: </b> <a class="tocSkip"></a>

Sucesso. Tudo está correto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário: </b> <a class="tocSkip"></a>

Observações. Algumas recomendações.
</div>

<div class="alert alert-block alert-danger">

<b>Comentário: </b> <a class="tocSkip"></a>

O bloco requer algumas correções. O trabalho não pode ser aceito com os comentários vermelhos.
</div>

Você pode me responder usando isto:

<div class="alert alert-block alert-info">
<b>Resposta do aluno</b> <a class="tocSkip"></a>
</div>



<div class="alert alert-block alert-danger">

<b>Comentário geral v1: </b> <a class="tocSkip"></a>


Parabéns pela conclusão do projeto, Micaela! 🎉 O projeto está muito bom, a maioria das tarefas foi realizadas corretamente!😊  

Para prosseguirmos com a aprovação, precisamos realizar ajustes no tópico:
    
           - Autor com a maior média de classificação (com pelo menos 50 classificações)
           - Número médio de avaliações de utilizadores com mais de 50 avaliações
    

    
Adicionei comentários para te ajudar na solução.

Estou à disposição para qualquer dúvida que você tiver! 

Boa sorte! 🍀
</div>

<div class="alert alert-block alert-success">

<b>Comentário geral v2: </b> <a class="tocSkip"></a>

Muuito obrigada pelos ajustes! Projeto aprovado!

# Projeto Final — Análise de Livros

Micaela

Este projeto tem como objetivo analisar uma base de dados de um serviço de livros, explorando informações sobre livros, autores, editoras e avaliações dos utilizadores.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-final-project-db'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode': 'require'})

In [2]:
pd.read_sql("SELECT 1", con=engine)

,?column?
0,1


## Teste de conexão

A conexão com o banco de dados foi estabelecida com sucesso, conforme demonstrado pela execução da query simples.

In [3]:
pd.read_sql("SELECT * FROM books LIMIT 5", con=engine)

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


## Exploração inicial dos dados

A tabela `books` contém informações sobre os livros disponíveis, incluindo título, número de páginas, data de publicação e editora.

É possível observar que existem livros com diferentes datas de publicação e volumes variados de páginas.

In [4]:
pd.read_sql("SELECT * FROM authors LIMIT 5", con=engine)

,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


A tabela `authors` contém os autores dos livros.

É possível observar que alguns registos incluem mais do que um autor no mesmo campo, separados por “/”, o que pode indicar coautoria.

In [5]:
pd.read_sql("SELECT * FROM publishers LIMIT 5", con=engine)

,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


A tabela `publishers` contém as editoras responsáveis pela publicação dos livros.

É possível observar variações nos nomes das editoras (por exemplo, "Ace", "Ace Book", "Ace Books"), o que pode indicar inconsistências ou diferentes divisões da mesma editora.

In [6]:
pd.read_sql("SELECT * FROM ratings LIMIT 5", con=engine)

,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


A tabela `ratings` contém as classificações atribuídas pelos utilizadores aos livros.

Cada registo representa uma avaliação individual, associando um utilizador a um livro e a uma pontuação, o que permite analisar a popularidade e qualidade percebida dos livros.

In [7]:
pd.read_sql("SELECT * FROM reviews LIMIT 5", con=engine)

,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


A tabela `reviews` contém os textos das avaliações feitas pelos utilizadores.

Ao contrário da tabela `ratings`, que contém apenas valores numéricos, esta tabela permite analisar opiniões mais detalhadas e qualitativas sobre os livros.

<div class="alert alert-block alert-success">

<b>Comentário: </b> <a class="tocSkip"></a>

Tabelas foram lidas corretamente

</div>



## Objetivo do estudo

O objetivo deste projeto é analisar os dados disponíveis para responder a questões relacionadas com a publicação de livros, avaliações dos utilizadores e desempenho de autores e editoras.

### 1. Quantos livros foram lançados depois de 1 de janeiro de 2000?

In [8]:
query = """
SELECT COUNT(*) AS total_books_after_2000
FROM books
WHERE publication_date > '2000-01-01'
"""
pd.read_sql(query, con=engine)

,total_books_after_2000
0,819


In [9]:
query = """
SELECT COUNT(*) AS total_books
FROM books
"""
pd.read_sql(query, con=engine)

,total_books
0,1000


Foram identificados 819 livros publicados após 1 de janeiro de 2000.

Ao comparar com o número total de livros na base de dados, é possível avaliar a proporção de obras recentes. Isso ajuda a perceber se o catálogo está mais focado em livros contemporâneos ou inclui uma grande quantidade de obras mais antigas.

A presença de um número significativo de livros publicados após 2000 pode indicar que a base de dados está relativamente atualizada e alinhada com tendências modernas de leitura.

### 2. Número de avaliações e classificação média por livro

In [10]:
query = """
SELECT
    b.book_id,
    b.title,
    COUNT(r.rating) AS total_ratings,
    AVG(r.rating) AS avg_rating
FROM books b
LEFT JOIN ratings r ON b.book_id = r.book_id
GROUP BY b.book_id, b.title
ORDER BY total_ratings DESC
"""
pd.read_sql(query, con=engine)

,book_id,title,total_ratings,avg_rating
0,948,Twilight (Twilight #1),160,3.662500
1,750,The Hobbit or There and Back Again,88,4.125000
2,673,The Catcher in the Rye,86,3.825581
3,75,Angels & Demons (Robert Langdon #1),84,3.678571
4,302,Harry Potter and the Prisoner of Azkaban (Harr...,82,4.414634
...,...,...,...,...
995,465,Naked Empire (Sword of Truth #8),2,3.500000
996,55,A Woman of Substance (Emma Harte Saga #1),2,5.000000
997,652,The Body in the Library (Miss Marple #3),2,4.500000
998,790,The Magicians' Guild (Black Magician Trilogy #1),2,3.500000


A análise mostra o número total de avaliações e a classificação média de cada livro.

Observa-se que alguns livros possuem um número elevado de avaliações, o que pode indicar maior popularidade ou visibilidade entre os utilizadores. Por outro lado, existem livros com poucas ou nenhuma avaliação, o que pode afetar a fiabilidade da sua classificação média.

A classificação média permite identificar a perceção geral dos leitores sobre cada livro. No entanto, é importante considerar que livros com poucas avaliações podem apresentar médias pouco representativas.

Assim, para uma análise mais robusta, seria recomendável considerar também o volume de avaliações ao interpretar a qualidade de cada livro.

### 3. Editora com mais livros com mais de 50 páginas

In [11]:
query = """
SELECT 
    p.publisher,
    COUNT(b.book_id) AS total_books
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY total_books DESC
LIMIT 1
"""
pd.read_sql(query, con=engine)

,publisher,total_books
0,Penguin Books,42


A análise identifica a editora que publicou o maior número de livros com mais de 50 páginas.

Este filtro permite excluir publicações muito curtas, focando apenas em livros mais completos. A editora apresentada no resultado é, assim, a que possui maior volume de publicações relevantes dentro deste critério.

### 4. Autor com maior média de classificação (mínimo 50 avaliações)

In [12]:
query = """
WITH book_ratings AS (
    SELECT
        b.book_id,
        b.author_id,
        COUNT(r.rating_id) AS num_ratings,
        AVG(r.rating) AS avg_rating
    FROM books AS b
    JOIN ratings AS r ON b.book_id = r.book_id
    GROUP BY b.book_id, b.author_id
    HAVING COUNT(r.rating_id) >= 50
),

author_avg_ratings AS (
    SELECT
        br.author_id,
        a.author,
        AVG(br.avg_rating) AS avg_author_rating
    FROM book_ratings AS br
    JOIN authors AS a ON br.author_id = a.author_id
    GROUP BY br.author_id, a.author
)

SELECT *
FROM author_avg_ratings
ORDER BY avg_author_rating DESC
LIMIT 1;
"""

pd.read_sql(query, con=engine)

,author_id,author,avg_author_rating
0,236,J.K. Rowling/Mary GrandPré,4.283844


<div class="alert alert-block alert-danger">
<b>Comentário: </b> <a class="tocSkip"></a>

A query acima necessita de ajustes.
    
Segue um exemplo de como resolver:
    

```python
    # Consulta para identificar o autor com a média mais alta de classificação de livros com pelo menos 50 classificações
    query = '''
    WITH book_ratings AS (
        SELECT 
            b.book_id,
            b.author_id,
            COUNT(r.rating_id) AS num_ratings,
            AVG(r.rating) AS avg_rating
        FROM books AS b
        JOIN ratings AS r ON b.book_id = r.book_id
        GROUP BY b.book_id, b.author_id
        HAVING COUNT(r.rating_id) >= 50
    ),
    author_avg_ratings AS (
        SELECT 
            br.author_id,
            a.author,
            AVG(br.avg_rating) AS avg_author_rating
        FROM book_ratings AS br
        JOIN authors AS a ON br.author_id = a.author_id
        GROUP BY br.author_id, a.author
    )
    SELECT 
    ... continue o código 
    LIMIT 1;
    '''
```
O output esperado é a autora J.K. Rowling/Mary GrandPré 
</div>


A análise identifica o autor com a maior classificação média considerando apenas livros com pelo menos 50 avaliações.

Este critério torna o resultado mais representativo, evitando médias influenciadas por poucos utilizadores.

### 5. Número médio de avaliações de utilizadores com mais de 50 avaliações

In [14]:
query = """
SELECT AVG(review_count) AS avg_reviews
FROM (
    SELECT
        username,
        COUNT(review_id) AS review_count
    FROM reviews
    WHERE username IN (
        SELECT
            username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    )
    GROUP BY username
) AS review_counts
"""

pd.read_sql(query, con=engine)

,avg_reviews
0,24.333333


<div class="alert alert-block alert-danger">

<b>Comentário: </b> <a class="tocSkip"></a>

O código necessita de ajustes.

Passo a passo para resolução:

1. Calcular a média da contagem de avaliações (reviews).
```sql
    SELECT AVG(review_count) AS avg_reviews
```

2. Subquery para contar as avaliações de texto (reviews) para usuários com mais de 50 avaliações.
 Exemplo: 
```sql
        SELECT 
            COUNT(*) AS review_count
        FROM
            reviews
        WHERE
            username IN (
                SELECT
                    username
                FROM
                    ratings
                GROUP BY
                    username
                HAVING
                    COUNT(rating_id) > 50
            )
        GROUP BY
            username
        ) AS review_counts
```
Para ajudar na validação da sua solução, **o resultado esperado é algo em torno de 24.**

</div>



A análise calcula o número médio de avaliações de texto feitas por utilizadores que realizaram mais de 50 classificações.

Este resultado permite compreender o comportamento dos utilizadores mais ativos da plataforma, indicando o nível médio de participação deste grupo.

## Conclusão

## Conclusão

A análise realizada permitiu responder às principais questões relacionadas com os livros, autores, editoras e comportamento dos utilizadores da plataforma.

Foi identificado um número significativo de livros publicados após o ano 2000, mostrando a presença de conteúdos relativamente recentes na base de dados. Também foi possível observar diferenças nas classificações médias dos livros e identificar a editora com maior número de publicações relevantes.

Além disso, a análise permitiu determinar o autor com melhor média de classificação considerando apenas livros com um mínimo de 50 avaliações, tornando o resultado mais representativo.

Por fim, verificou-se o nível médio de participação dos utilizadores mais ativos da plataforma, com base no número de avaliações de texto realizadas.

No geral, os resultados ajudam a compreender melhor a dinâmica da plataforma e o comportamento dos seus utilizadores.

link do pdf da apresentação: https://drive.google.com/drive/folders/18_XqLTn07d4mLy4_S_AsD2-34dCLiRCX?usp=sharing

<div class="alert alert-block alert-success">

<b>Comentário: </b> <a class="tocSkip"></a>

A-DO-REI a apresentação :) Belo gosto!!